### PR TITLE
improved sitemap and robots.txt files

### PR DESCRIPTION
**Pull request by vinothkannans**
updates & news features from @SamSaffron's [comment](https://meta.discourse.org/t/discourse-sitemap-plugin/40348/32?u=vinothkannans)

> 1. There is no built in extensibility to robots.txt generation, this needs to be added so it does not overwrite it completely (instead plugin should inject the instructions). Otherwise, each time we touch robots.txt plugin needs to be updated.
> 2. The way it iterates using the object model is inefficient for so many records, instead this should be carefully plucking the data it needs or even possibly using raw sql.
> 3. Nothing is cached which is risky, instead ... something should be cached. Perhaps site map should be re-generated daily or something along those lines.
> 4. Sitemap is most useful for gigantic sites, it should use a sitemap index and multiple sitemap files (splitting site into 50k chunks)

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -20,63 +20,80 @@ class Engine < ::Rails::Engine
   require_dependency "application_controller"
 
   class DiscourseSitemap::SitemapController < ::ApplicationController
-    def generate
-      prepend_view_path "plugins/discourse-sitemapcrap/app/views/" 
-      # this code is never called somehow
-    end
-  end
+    layout false
+    skip_before_filter :preload_json, :check_xhr, :redirect_to_login_if_required
```
Comment (SamSaffron) [inline plugin.rb]: I would not skip redirect_to_login_if_required

Comment (vinothkannans) [reply plugin.rb]: [RobotsTxtController](https://github.com/discourse/discourse/blob/master/app/controllers/robots_txt_controller.rb#L3) skipping the `redirect_to_login_if_required` filter. And `sitemap.xml` file's reference included in `robots.txt` file. So after looking into `robots.txt` file automatically bots will be looking for `sitemap.xml` file I guess.

In this case still I can remove it? I am not sure about how bot things will work.

**Comment by SamSaffron** [comment]
I would remove, sitemap actually exposes information, robots does not.

Also, if you require login I would not even inject the sitemap into robots,
pointless.

**Comment by vinothkannans** [comment]
> I would remove, sitemap actually exposes information, robots does not.

Yes. It should be. Very good explanation.

> if you require login I would not even inject the sitemap into robots, pointless.

I got it. I will do the implementation like that :+1:

**Comment by communiteq** [comment]
Thanks a million @vinothkannans  and @SamSaffron !

**Pull request merged**